### PR TITLE
fix: 修复 chunk_bwd_dv_local 特定 GQA 配置 kernel hang

### LINKED ISSUE #462: [Bug]: chunk_bwd_dv_local 在特定 GQA head 配置下触发 NPU kernel hang
### 环境信息

python version: 3.9.25
CANN version: version=26.1.0
CANN public version: 9.2.0
torch version: 2.7.1+cpu
torch_npu version: 2.7.1.post6.dev20260723
triton-ascend version: 3.2.0
ascend：910B

### 最小复现

```python
import math
import torch
from fla_npu.ops import ascendc as ascendc_ops

torch.npu.config.allow_internal_format = False
torch.npu.set_compile_mode(jit_compile=False)
torch.npu.set_device(0)

B, HK, HV, T, K, V, chunk_size, scale = 1, 15, 120, 1, 128, 128, 64, 1.0 / math.sqrt(K)

torch.manual_seed(1080)
q = (torch.rand(B, HK, T, K) * 2 - 1).to(torch.float16).npu()
k = (torch.rand(B, HK, T, K) * 2 - 1).to(torch.float16).npu()
d_o = (torch.rand(B, HV, T, V) * 2 - 1).to(torch.float16).npu()
g = (-torch.rand(B, HV, T)).to(torch.float32).npu()

print("NPU BEFORE OP", flush=True)
dv = ascendc_ops.npu_chunk_bwd_dv_local(
    q,
    k,
    d_o,
    g,
    scale=scale,
    chunk_size=chunk_size,
    g_gamma=None,
    A=None,
    cu_seqlens=None,
    chunk_indices=None,
)
print("NPU AFTER OP", flush=True)

torch.npu.synchronize()
print("NPU AFTER SYNC", flush=True)


复现参数：


q:             fp16 [1, 15, 1, 128]
k:             fp16 [1, 15, 1, 128]
d_o:           fp16 [1, 120, 1, 128]
g:             fp32 [1, 120, 1]
scale:         1 / sqrt(128)
chunk_size:    64
g_gamma:       None
A:             None
cu_seqlens:    None
chunk_indices: None
```

### 完整报错

```text
单独运行上述 case 时没有立即产生 Python exception，而是稳定卡在：

torch.npu.synchronize()

日志表现为：

NPU BEFORE OP
NPU AFTER OP

但 `torch.npu.synchronize()` 无法返回，因此不会打印：

NPU AFTER SYNC
```

### 期望行为

根据当前接口约束，`HV % HK == 0` 的 GQA head 配置应能够正常执行。

如果该 head 配置实际上不在算子支持范围内，也期望在参数校验或 tiling 阶段明确返回错误，而不是 kernel launch 后一直无法完成，导致 NPU stream 和后续测试任务被阻塞。

### 其他补充

对 `HK`、`HV` 进行了单变量对照测试，结果如下：

| HK | HV  | HV / HK | 结果   |
| -- | --- | ------- | ---- |
| 2  | 16  | 8       | PASS |
| 14 | 112 | 8       | PASS |
| 15 | 120 | 8       | HANG |
| 16 | 128 | 8       | HANG |
| 15 | 15  | 1       | PASS |
| 15 | 30  | 2       | PASS |
| 15 | 60  | 4       | PASS |
| 30 | 120 | 4       | PASS |

### 其他补充

_No response_

### PR DESCRIPTION
**Pull request by chen-linxin**
## 关联 Issue / 背景

- 关联 Issue: Fixes #462
- 背景与目标: 修复 `chunk_bwd_dv_local` 在特定 GQA head 配置下发生的 kernel hang。
- 本 PR 解决的问题: `H_qk=15`、`H_do=120`、`hRatio=8` 等配置下，`torch.npu.synchronize()` 无法返回。

## 变更类型

- [x] Bug 修复
- [x] 算子精度 / 稳定性修复
- [x] 文档 / 示例 / 测试更新

## 算子责任范围

- 涉及算子名称: `chunk_bwd_dv_local`
- 修改范围:
  - [x] `op_kernel` / Ascend C Kernel
  - [x] 单算子测试
  - [x] 文档
- 输入 / 输出 / shape / dtype / format 支持范围: 不变。
- 明确不包含的范围: 不修改 tiling、算子接口、Torch 适配或计算公式。
- 是否影响公共模块或其他算子: 否。
- ABI / 接口兼容性:
  - [x] 不涉及算子 `def`、`aclnn` 接口或 `torch` 接口入参、返回值、必选可选属性变化

## 计算位置与流水

对每个 `doHead=j`，令 `r=H_do/H_qk`、`h=floor(j/r)`：

$$S_h=K_hQ_h^T$$

$$P_j=scale\cdot M\odot\exp(g_{j,col}-g_{j,row})\odot S_h$$

$$dV_j=P_jdO_j$$

```mermaid
flowchart LR
    KQ["Q / K"] --> P1["Phase 1 · AIC<br/>S_h = K_h Q_h^T"]
    P1 -->|"QK_READY"| QKS["QK workspace<br/>3 个 ring slot"]
    QKS --> P15["Phase 1.5 · AIV<br/>P_j = scale · mask(exp(g_col-g_row)) ⊙ S_h"]
    P15 -->|"GATED_READY"| GS["gated workspace"]
    GS --> P2["Phase 2 · AIC<br/>dV_j = P_j dO_j"]
    P2 --> DV["dV"]

    QKS -. "QK ready/free 同步边界" .-> P15
    GS -. "gated ready/free 同步边界" .-> P2
```

`QK ready/free` 位于 $S_h$ 生成后、AIV 使用 $S_h$ 计算 $P_j$ 前；`gated ready/free` 位于 $P_j$ 写回后、AIC 使用 $P_j$ 计算 $dV_j$ 前。问题发生在这两个计算阶段之间的同步边界，而不是数值公式本身。

## 问题原因

QK workspace 只有 3 个 ring slot，但旧实现的 QK ready/free 和 gated ready/free 均采用每 15 次批量反向确认的同步协议。每个 QK head 会派生 `hRatio` 个 gated head，因此两条同步链推进速度不同。

在 Issue 配置中，AIC 发送 `qkHead=14` 时到达第 15 个 `QK_READY`，开始等待 QK 的批量反向确认。受 3-slot 流水滞后影响，此时 AIC 只消费到 `qkHead=11`，即 96 个 gated 事件。AIV 要消费第 15 个 QK ready，必须先处理 `qkHead=12/13`；但它在 `qkHead=13` 的第一个输出处到达第 105 个 gated 事件，并开始等待 gated 的批量反向确认，最终形成循环等待。

```mermaid
flowchart TD
    A["AIC 已发送第 15 个 QK_READY<br/>对应 qkHead=14"]
    B["AIC 等待批量 QK_FREE 确认"]
    C["AIV 必须消费第 15 个 QK_READY<br/>才能返回该确认"]
    D["AIV 处理 qkHead=13 时<br/>到达第 105 个 GATED_READY"]
    E["AIV 等待批量 GATED_FREE 确认"]
    F["AIC 必须继续消费第 105 个<br/>GATED_READY 才能返回确认"]

    A --> B
    B --> C
    C --> D
    D --> E
    E --> F
    F -->|"但 AIC 已阻塞在 QK_FREE"| B
```

## 修改内容

- 将 QK ready/free 改为逐 QK head 的 `CrossCoreFlag` 信用同步。
- AIV 将当前 QK 数据搬入 UB 队列后立即返回 QK free，解除 QK ready 的计数压力。
- gated ready/free 保持不变；AIC 仍在全部 `hRatio` 个 gated workspace 写回后才复用对应 ring slot，因此不会引入 workspace 复用竞态。
- 同步修改普通 Vector 和 Ascend 950 arch35 实现。
- Fast-Kernel 固定长度测试增加 Issue 原始 GQA 配置。
- README 补充 QK 流水和核间同步说明。

```mermaid
sequenceDiagram
    participant AIC
    participant GM as 3-slot workspace
    participant AIV

    AIC->>GM: 写入 S_h = K_h Q_h^T
    AIC->>AIV: QK_READY(h)
    AIV->>GM: 将 S_h 搬入 UB 队列
    AIV->>AIC: 立即返回 QK_FREE(h)

    loop hRatio 个 dO head
        AIV->>GM: 写入 P_j
        AIV->>AIC: GATED_READY(j)
    end

    AIC->>GM: 读取全部 P_j
    AIC->>AIC: 计算 dV_j = P_j dO_j
    Note over AIC,GM: gated-ready 完成后才复用对应 ring slot
```

### 编译矩阵

| 产品 | `--soc` | CANN 版本 | 结果 |
| --- | --- | --- | --- |
| A2 | `ascend910b` | 9.2 | 单算子包和 wheel 编译通过 |
| A3 | `ascend910_93` | 9.2 | 单算子包交叉编译通过 |
| A5 | `ascend950` | 9.2 | 单算子包交叉编译通过 |

### A2 运行验证

- Fast-Kernel 测试 14/14 通过。
- Issue 原始配置通过，最大绝对误差 `0.001953`。
- `H_qk=17`、`H_do=119`、`hRatio=7` 补充配置通过，最大绝对误差 `0.007780`。
- Example ST 前向及 `dq/dk/dv/dbeta/dg` 精度检查全部通过。
- sanitizer synccheck 未发现 ERROR、死锁或超时；仅报告公共 GEMM 模板已有的冗余 `set_flag` 警告。
- sanitizer racecheck 报告 `No error detected`。
- 临界场景性能中位数：修改前 `0.063313 ms`，修改后 `0.060543 ms`，未观察到性能回退。
- A3/A5 本轮仅完成交叉编译，未执行硬件运行测试。

## 兼容性、风险与回退

- 兼容性说明: 接口、shape、dtype、format、tiling 和 ABI 均未变化。
- 风险点: QK free 的发送时机发生变化；gated-ready 仍负责保证 ring slot 的安全复用。
- 回退方案: 回退本 PR 即可恢复原同步协议。
- 回退责任确认:
  - [x] 若本 PR 未满足编译 / 测试要求，或引入阻塞性 bug，PR 提交人承诺第一时间回退或配合维护者完成回退
  - [x] 回退后会同步说明影响范围、恢复版本、后续修复计划

## Checklist

- [x] 已关联 Issue
- [x] 已明确本 PR 的算子责任范围和不包含范围
- [x] CANN 9.0 及以上已验证 A2 / A3 / A5 编译通过
- [ ] 已验证修改算子的对应 test 在 A2 / A3 / A5 运行通过（A2 已通过；A3/A5 仅交叉编译）
- [ ] 已验证整体 Example ST 在 A2 / A3 / A5 通过（A2 已通过；A3/A5 未执行）
- [x] 已完成必要的精度、性能和同步回归验证
- [x] 已确认没有引入无关格式化或大规模重构
- [x] 已更新相关 README 和回归测试
- [x] PR 标题已使用 `fix:` 类型标签
- [x] 已确认如不满足准入要求或引入 bug，将第一时间回退

### 最新精度标准回归

- 按仓库最新精度测试标准执行：`mixed_tolerance_bm`、CPU FP64 标杆、GM 初始化，每个独立精度场景至少使用 3 个固定 seed，并运行全部 `(case, seed)` 组合。
- 200 个基础场景扩展为 600 个三 seed 组合，并增加 3 个 `H_qk=15`、`H_do=120`、`hRatio=8` 的 Issue #462 定向回归组合，共 603 个精度用例。
- 覆盖 BF16/FP16、`g` 的 BF16/FP16/FP32、`V=128/256`、`chunk_size=64/128`、定长及变长接口路径。
- executor 根据 `case_spec` 实际构造 `g_dtype` 和变长元数据，CPU 标杆与 NPU DUT 使用相同场景。
- 测试 JSON 与 README 已同步更新。

验证结果：

- A2、ATK 26.8.8：603/603 执行成功，执行失败 0。
- 精度 603/603 通过，通过率 100%，最终结果 `Pass`。
- `hRatio=8` 定向回归 3/3 通过，未出现 kernel hang。

### 其他已完成验证

- 确定性 16/16 通过，每条循环 50 次。
- 内存检测 16/16 通过，目标算子区间非法访问为 0。
- 性能测试 8/8 完成，平均耗时 `4.438 us`。

### CODE REVIEW

**Comment by weinachuan** [comment]
【review】
问题分类：编码红线问题
位置：fla/ops/ascendc/gdn/chunk_gdn_bwd/chunk_bwd_dv_local/op_kernel/chunk_bwd_dv_local_cube.h:170，代码：CrossCoreWaitFlag(aivToAicQkFreeFlag)
问题：每个 qkHead 的两个 AIV 子核都会在各自的 Vector 实例中执行 CrossCoreSetFlag(aivToAicQkFreeFlag)，但 AIC 对同一 qkHead 只执行一次 WaitFlag。CrossCore flag 的一次 set 是一个独立 credit，不会自动合并为“两子核共同归还的一次”；因此从第一个 head 起就会遗留一个 free credit。
可能导致的结果：遗留 credit 会让后续 qkHead 的 wait 提前通过。3-slot ring 回绕时，AIC 可能在另一个 AIV 仍从该 slot 的 GM 搬运 QK 数据前复写 slot，导致数据竞争、精度错误或新的卡死。当前新增用例只验证最终完成，未覆盖两个 AIV 对同一 slot 的 free 全部被消费的配对关系。
建议的修改方法：为每个 qkHead 消费两个 AIV free credit，或改为只有在两个 AIV 都完成 QK 搬入后才产生一份聚合 free；同时补充显式覆盖 1C2V、slot 回绕和两子核完成时序错开的同步测试。

**Comment by chen-linxin** [comment]
@weinachuan 感谢检视。这里使用的是 `CrossCoreSetFlag<0x2>` 的 Mode 2，两个 AIV 的 set 不会分别在 AIC 侧留下独立 credit。

Mode 2 的配对语义是：

- AIC 执行一次 set 后，AIV0、AIV1 各自的计数器增加 1，两个 AIV 分别执行一次 wait。
- AIV 反向通知 AIC 时，只有 AIV0、AIV1 都执行 set 后，AIC 侧计数器才增加 1。因此两个 AIV 的 set 共同对应 AIC 的一次 wait。

当前 kernel 配置为 `KERNEL_TYPE_MIX_AIC_1_2`。两个 AIV 都在各自的 QK MTE2 搬运进入 UB 队列后执行 `CrossCoreSetFlag<0x2, PIPE_MTE2>`，所以 AIC 的 `CrossCoreWaitFlag` 只有在两个 AIV 都完成搬运后才会通过，不会残留 free credit，也不会因其中一个 AIV 提前完成而提前复用 slot。后续 `gated-ready` 仍会等全部 `hRatio` 个输出写回后再允许对应 ring slot 被覆盖。

Catlass 当前固定版本的封装只是将 Mode 和 flagId 原样转发给 Ascend C API，没有改变上述计数语义。如果在 AIC 侧增加第二次 wait，由于两个 AIV 只共同产生一份聚合 credit，第二次 wait 反而会阻塞。

官方 Mode 2 定义：https://www.hiascend.com/document/detail/zh/canncommercial/83RC1/API/ascendcopapi/atlasascendc_api_07_0273.html

另外，新增的 `T=1、H_qk=15、H_do=120、hRatio=8` 回归中，两个 AIV 的任务量天然不一致，并包含 ring slot 多次回绕；3 个固定 seed 均已通过，未出现精度错误或卡死。因此这里不需要按“两份独立 credit”修改。
